hpx/runtime/threads/thread_id_type.hpp

```diff
@@ -0,0 +1,142 @@
+//  Copyright (c) 2018 Thomas Heller
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+/// \file hpx/runtime/threads/thread_id_type.hpp
+
+#ifndef HPX_THREADS_THREAD_ID_TYPE_HPP
+#define HPX_THREADS_THREAD_ID_TYPE_HPP
+
+#include <cstddef>
+#include <functional>
+
+namespace hpx { namespace threads
+{
+    class HPX_EXPORT thread_data;
+
+    struct thread_id_type
+    {
+        constexpr thread_id_type()
+          : thrd_(nullptr) {}
+        constexpr thread_id_type(thread_data* thrd)
```

Comment:
We should make this constructor `explicit` to avoid unexpected conversions (we already have an implicit operator `thread_data*()` below).